shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/ClangArgs.kt

```diff
@@ -37,146 +37,174 @@ class ClangArgs(private val configurables: Configurables) : Configurables by con
         KonanTarget.MACOS_X64 -> "$absoluteTargetToolchain/usr/bin"
         else -> throw TargetSupportException("Unexpected host platform")
     }
 
-    private val extraHostClangArgs =
-            if (configurables is GccConfigurables) {
-                listOf("--gcc-toolchain=${configurables.absoluteGccToolchain}")
-            } else {
-                emptyList()
+    // TODO: Use buildList
+    private val commonClangArgs: List<String> = mutableListOf<List<String>>().apply {
+        add(listOf("-B$binDir", "-fno-stack-protector"))
+        if (configurables is GccConfigurables) {
+            add(listOf("--gcc-toolchain=${configurables.absoluteGccToolchain}"))
+        }
+        if (configurables is TargetableConfigurables) {
+            add(listOf("-target", configurables.targetArg!!))
+        }
+    }.flatten()
+
+    private val osVersionMin: String
+            get() {
+                require(configurables is AppleConfigurables)
+                return configurables.osVersionMin
             }
 
-    val commonClangArgs = listOf("-B$binDir", "-fno-stack-protector") + extraHostClangArgs
+    val specificClangArgs: List<String> = when (target) {
+        KonanTarget.LINUX_X64 -> listOf(
+                "--sysroot=$absoluteTargetSysRoot",
+        )
+
+        KonanTarget.LINUX_ARM32_HFP -> listOf(
+                "-mfpu=vfp", "-mfloat-abi=hard",
+                "--sysroot=$absoluteTargetSysRoot",
+                // TODO: those two are hacks.
+                "-I$absoluteTargetSysRoot/usr/include/c++/4.8.3",
+                "-I$absoluteTargetSysRoot/usr/include/c++/4.8.3/arm-linux-gnueabihf"
+        )
+
+        KonanTarget.LINUX_ARM64 -> listOf(
+                "--sysroot=$absoluteTargetSysRoot",
+                "-I$absoluteTargetSysRoot/usr/include/c++/7",
+                "-I$absoluteTargetSysRoot/usr/include/c++/7/aarch64-linux-gnu"
+        )
+
+        KonanTarget.LINUX_MIPS32 -> listOf(
+                "--sysroot=$absoluteTargetSysRoot",
+                "-I$absoluteTargetSysRoot/usr/include/c++/4.9.4",
+                "-I$absoluteTargetSysRoot/usr/include/c++/4.9.4/mips-unknown-linux-gnu"
+        )
+
+        KonanTarget.LINUX_MIPSEL32 -> listOf(
+                "--sysroot=$absoluteTargetSysRoot",
+                "-I$absoluteTargetSysRoot/usr/include/c++/4.9.4",
+                "-I$absoluteTargetSysRoot/usr/include/c++/4.9.4/mipsel-unknown-linux-gnu"
+        )
+
+        KonanTarget.MINGW_X64, KonanTarget.MINGW_X86 -> listOf(
+                "--sysroot=$absoluteTargetSysRoot", "-Xclang",
+                "-flto-visibility-public-std"
+        )
+
+        KonanTarget.MACOS_X64 -> listOf(
+                "--sysroot=$absoluteTargetSysRoot",
+                "-mmacosx-version-min=$osVersionMin"
+        )
+
+        KonanTarget.IOS_ARM32 -> listOf(
+                "-stdlib=libc++",
+                "-arch", "armv7",
+                "--sysroot=$absoluteTargetSysRoot",
```

Comment:
Are you absolutely sure that `--sysroot` is equivalent to `-isysroot`?

Reply:
AFAIU, yes. But the devil is in the details, I'l recheck Clang source.

Reply:
If I remember right on Darwin platforms `-isysroot` is exactly the same as `--sysroot`, in other there are some changes.